main_test.go

```diff
@@ -379,72 +377,148 @@ func runTest(t *testing.T, test *mainTest) {
-
-	filenameEscapeRE := regexp.MustCompile(`%[0-9A-Fa-f]{2}`)
-
-	for _, input := range match {
-		file := filenameEscapeRE.ReplaceAllStringFunc(input, func(s string) string {
-			code, err := strconv.ParseUint(s[1:], 16, 8)
-			if err != nil {
-				panic(err)
-			}
-			return string([]byte{byte(code)})
-		})
-		if file != input {
-			if err := os.Link(input, file); err != nil {
-				if !os.IsExist(err) {
-					return fmt.Errorf("linking %s -> %s: %v", file, input, err)
-				}
-			}
-			t.Logf("Linked %s -> %s", input, file)
+	tests := make([]mainTest, 0, len(files))
+	for _, input := range files {
+		name := strings.TrimSuffix(input, ".jsonnet")
+		var meta testMetadata
+		if val, exists := metadataForTests[name]; exists {
+			meta = val
 		}
+		tests = append(tests, mainTest{name: name, input: input, golden: name + ".golden", meta: &meta})
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			runTest(t, &test)
+		})
 	}
-
-	return nil
 }
 
-func TestMain(t *testing.T) {
-	flag.Parse()
-
-	if err := expandEscapedFilenames(t); err != nil {
+func withinWorkingDirectory(t *testing.T, dir string) func() error {
+	t.Helper()
+	cwd, err := os.Getwd()
+	if err != nil {
 		t.Fatal(err)
 	}
-
-	var mainTests []mainTest
-	match, err := filepath.Glob("testdata/*.jsonnet")
-	if err != nil {
+	if err := os.Chdir(dir); err != nil {
 		t.Fatal(err)
 	}
+	return func() error {
+		return os.Chdir(cwd)
+	}
+}
 
-	jsonnetExtRE := regexp.MustCompile(`\.jsonnet$`)
+func TestEvalUnusualFilenames(t *testing.T) {
+	// Escaped filenames exist because their unescaped forms are invalid on Windows. We have no
+	// choice but to skip these in testing.
+	if runtime.GOOS == "windows" {
+		return
+	}
 
-	for _, input := range match {
-		// Skip escaped filenames.
-		if strings.ContainsRune(input, '%') {
-			continue
+	// Are we running within "bazel test"?
+	dir := os.Getenv("TEST_TMPDIR")
+	if len(dir) == 0 {
+		var err error
+		if dir, err = ioutil.TempDir("", "jsonnet"); err != nil {
+			t.Fatal(err)
 		}
-		name := jsonnetExtRE.ReplaceAllString(input, "")
-		golden := jsonnetExtRE.ReplaceAllString(input, ".golden")
-		var meta testMetadata
-		if val, exists := metadataForTests[name]; exists {
-			meta = val
+	}
+
+	copySmallFile := func(t *testing.T, dst, src string) {
+		b, err := ioutil.ReadFile(src)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if err := ioutil.WriteFile(dst, b, 0444); err != nil {
+			t.Fatal(err)
 		}
-		mainTests = append(mainTests, mainTest{name: name, input: input, golden: golden, meta: &meta})
 	}
-	for _, test := range mainTests {
-		t.Run(test.name, func(t *testing.T) {
-			runTest(t, &test)
+	// These files are imported by files below, but we don't need to exercise their round-trip
+	// behavior here, as they're covered by TestEval above.
+	for _, f := range []string{"true"} {
+		for _, ext := range []string{".jsonnet"} {
+			name := f + ext
+			copySmallFile(t, filepath.Join(dir, name), filepath.Join("testdata", name))
```

Comment:
Note that I considered trying to run these evaluation tests using just in-memory content, avoiding the filesystem, but two things motivated doing it with files instead:
- We can reuse the `runTest` function, which assumes that it will use files as input.
- We'd have to rig up our own `Importer` to handle the test cases where the top-level input file imports another one of these files.  
  Writing just the _true.jsonnet_ file probably would have been sufficient to cover this need, but the first motivation argued against duplicating or factoring the common parts from the `runTests` function.

Let me know if you'd prefer that we just process in-memory content here.